### PR TITLE
Adds support for AVLdiy.cn D512 usb dmx device, closes #1073

### LINKED ISSUE #1073: add support for AVLdiy.cn device
It is basically the same as Anymau
All I did was change 
```c
const char AnymauDMXFactory::EXPECTED_MANUFACTURER[] = "AVLdiy.cn";
const char AnymauDMXFactory::EXPECTED_PRODUCT[] = "D512";
const uint16_t AnymauDMXFactory::PRODUCT_ID = 0x8888;
const uint16_t AnymauDMXFactory::VENDOR_ID = 0x03EB;
```
and udev rules
```
SUBSYSTEM=="usb*", ACTION=="add", ATTRS{idVendor}=="03eb", ATTRS{idProduct}=="8888", MODE="0666", GROUP="plugdev"
SUBSYSTEM=="usb*", ACTION=="add", SYSFS{idVendor}=="03eb", SYSFS{idProduct}=="8888", MODE="0666", GROUP="plugdev"
```
and worked

### PR DESCRIPTION
**Pull request by thgjerde**
The AVLdiy.cn D512 uses the same protocol as the Anyma uDMX device.

### CODE REVIEW

File: plugins/usbdmx/AVLdiyD512.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+class AVLdiyD512: public SimpleWidget {
+ public:
+  /**
+   * @brief Create a new AVLdiyD512.
+   * @param adaptor the LibUsbAdaptor to use.
+   * @param usb_device the libusb_device to use for the widget.
+   * @param serial the serial number of the widget.
+   */
+  AVLdiyD512(ola::usb::LibUsbAdaptor *adaptor,
+            libusb_device *usb_device,
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Indent by one to match the (

File: plugins/usbdmx/AVLdiyD512.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            const std::string &serial)
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Indent by one to match the (

File: plugins/usbdmx/AVLdiyD512.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+class SynchronousAVLdiyD512: public AVLdiyD512 {
+ public:
+  /**
+   * @brief Create a new SynchronousAVLdiyD512.
+   * @param adaptor the LibUsbAdaptor to use.
+   * @param usb_device the libusb_device to use for the widget.
+   * @param serial the serial number of the widget.
+   */
+  SynchronousAVLdiyD512(ola::usb::LibUsbAdaptor *adaptor,
+                       libusb_device *usb_device,
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Indent by one to match the (

File: plugins/usbdmx/AVLdiyD512.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                       const std::string &serial);
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Indent by one to match the (

File: plugins/usbdmx/AVLdiyD512.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+class AsynchronousAVLdiyD512 : public AVLdiyD512 {
+ public:
+  /**
+   * @brief Create a new AsynchronousAVLdiyD512.
+   * @param adaptor the LibUsbAdaptor to use.
+   * @param usb_device the libusb_device to use for the widget.
+   * @param serial the serial number of the widget.
+   */
+  AsynchronousAVLdiyD512(ola::usb::LibUsbAdaptor *adaptor,
+                        libusb_device *usb_device,
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Indent by one to match the (

File: plugins/usbdmx/AVLdiyD512.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                        const std::string &serial);
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Indent by one to match the (

File: debian/ola.udev [hunk 7]
```diff
@@ -4,6 +4,9 @@ SUBSYSTEM=="usb|usb_device", ACTION=="add", ATTRS{idVendor}=="0403", ATTRS{idPro
 # udev rules for the anyma dmx device
 SUBSYSTEM=="usb|usb_device", ACTION=="add", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="05dc", GROUP="plugdev"
 
+# udev rules for the avldiy dmx device
```
Comment (peternewman) [inline debian/ola.udev]: Can you also open a PR here to patch the pi file: https://github.com/OpenLightingProject/raspberrypi/blob/master/etc/udev/rules.d/10-local.rules

File: plugins/usbdmx/AVLdiyD512.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * AVLdiyD512.cpp
+ * The synchronous and asynchronous AVLdiy D512 widgets.
+ * Copyright (C) 2014 Simon Newton
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.cpp]: Correct this to your name and 2016.

File: plugins/usbdmx/AVLdiyD512.h [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * AVLdiyD512.h
+ * The synchronous and asynchronous AVLdiy D512 widgets.
+ * Copyright (C) 2014 Simon Newton
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Correct this to your name and 2016.

File: plugins/usbdmx/AVLdiyD512Factory.cpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * AVLdiyD512Factory.cpp
+ * The factory for AVLdiy D512 widgets.
+ * Copyright (C) 2014 Simon Newton
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512Factory.cpp]: Correct this to your name and 2016.

File: plugins/usbdmx/AVLdiyD512Factory.h [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * AVLdiyD512Factory.h
+ * The factory for AVLdiy D512 widgets.
+ * Copyright (C) 2014 Simon Newton
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512Factory.h]: Correct this to your name and 2016.

**Comment by peternewman** [comment]
If you resync @thgjerde the Mac tests should now pass.

**Comment by peternewman** [comment]
This will close https://github.com/OpenLightingProject/ola/issues/1073 .

**Comment by thgjerde** [comment]
I don't see an easy way to inherit from the AnymauDMX classes without refactoring all of the SimpleWidget inheritance tree.

**Comment by peternewman** [comment]
Okay, I'd not really looked, just throwing it out there.

Can you avoid squashing in future please, as it makes my job harder to see if you've fixed the comments.

File: plugins/usbdmx/AVLdiyD512.cpp (hunk 8 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.cpp]: Done.

File: plugins/usbdmx/AVLdiyD512Factory.cpp (hunk 10 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512Factory.cpp]: Done.

File: plugins/usbdmx/AVLdiyD512.h (hunk 1 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Done.

File: plugins/usbdmx/AVLdiyD512.h (hunk 2 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Done.

File: plugins/usbdmx/AVLdiyD512.h (hunk 4 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Done.

File: plugins/usbdmx/AVLdiyD512.h (hunk 5 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Done.

File: plugins/usbdmx/AVLdiyD512.h (hunk 6 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.h]: Done.

File: debian/ola.udev (hunk 7 above)
Comment (thgjerde) [inline debian/ola.udev]: https://github.com/OpenLightingProject/raspberrypi/pull/8

**Comment by peternewman** [comment]
Can you also add a mention of your interface to: https://github.com/OpenLightingProject/ola/blob/master/plugins/usbdmx/UsbDmxPlugin.cpp#L95

You may also want to add yourself here:
https://github.com/OpenLightingProject/ola/blob/master/AUTHORS

File: plugins/usbdmx/AVLdiyD512.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+ private:
+  LibUsbAdaptor* const m_adaptor;
+
+  bool TransmitBuffer(libusb_device_handle *handle,
+                      const DmxBuffer &buffer);
+};
+
+bool AVLdiyThreadedSender::TransmitBuffer(libusb_device_handle *handle,
+                                         const DmxBuffer &buffer) {
```
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.cpp]: Indent by one to match the (

File: plugins/usbdmx/SyncPluginImpl.cpp [hunk 13]
```diff
@@ -116,6 +119,13 @@ bool SyncPluginImpl::Stop() {
   return true;
 }
 
+bool SyncPluginImpl::NewWidget(AVLdiyD512 *widget) {
```
Comment (peternewman) [inline plugins/usbdmx/SyncPluginImpl.cpp]: Can you put this below the AnymauDMX method.

File: plugins/usbdmx/SyncPluginImpl.cpp [hunk 14]
```diff
@@ -66,6 +68,7 @@ SyncPluginImpl::SyncPluginImpl(PluginAdaptor *plugin_adaptor,
       m_debug_level(debug_level),
       m_preferences(preferences),
       m_context(NULL) {
+  m_widget_factories.push_back(new AVLdiyD512Factory(&m_usb_adaptor));
```
Comment (peternewman) [inline plugins/usbdmx/SyncPluginImpl.cpp]: Can you put this below the AnymauDMX method.

**Comment by peternewman** [comment]
Aside from my final comments, anything else from yourself @nomis52 ?

**Comment by thgjerde** [comment]
I think I have addressed all the requested changes.

**Comment by thgjerde** [comment]
Please let me know if there is anything more I should fix.

**Comment by peternewman** [comment]
Thanks @thgjerde , sorry I've been away, I'll take a proper look tonight and should be able to merge them from the looks of things.

File: plugins/usbdmx/AVLdiyD512.cpp (hunk 12 above)
Comment (peternewman) [inline plugins/usbdmx/AVLdiyD512.cpp]: Done.

File: plugins/usbdmx/SyncPluginImpl.cpp (hunk 14 above)
Comment (peternewman) [inline plugins/usbdmx/SyncPluginImpl.cpp]: Done.

File: plugins/usbdmx/SyncPluginImpl.cpp (hunk 13 above)
Comment (peternewman) [inline plugins/usbdmx/SyncPluginImpl.cpp]: Done.

**Comment by peternewman** [comment]
LGTM

**Pull request merged**